File: sdks/php/src/Node/NodeHandler.php
<?php

declare(strict_types=1);

namespace Blok\Blok\Node;

use Blok\Blok\Types\Context;

/**
 * NodeHandler is the core interface that all Blok nodes must implement.
 *
 * Example:
 * ```php
 * class MyNode implements NodeHandler
 * {
 *     public function execute(Context $ctx, array $config): mixed
 *     {
 *         $name = $ctx->request->bodyStr('name') ?? 'World';
 *         return ['message' => "Hello, {$name}!"];
 *     }
 * }
 * ```
 */
interface NodeHandler
{
    /**
     * Execute the node logic with the given workflow context and node configuration.
     *
     * @param Context $ctx The workflow execution context
     * @param array<string, mixed> $config Node-specific configuration
     * @return mixed The result data to include in the ExecutionResult
     * @throws \Throwable On execution failure
     */
    public function execute(Context $ctx, array $config): mixed;
}

File: sdks/ruby/lib/blok/node/node_handler.rb
# frozen_string_literal: true

module Blok
  module Node
    # NodeHandler is the base class that all Blok nodes must inherit from.
    #
    # Subclasses must implement the +execute+ method which receives the
    # workflow context and node-specific configuration, and returns a
    # data hash (or any JSON-serializable value).
    #
    # @example
    #   class GreetNode < Blok::Node::NodeHandler
    #     def execute(ctx, config)
    #       name = ctx.request.body_str("name") || "World"
    #       prefix = config["prefix"] || "Hello"
    #       { "message" => "#{prefix}, #{name}!" }
    #     end
    #   end
    #
    class NodeHandler
      # Subclasses defined anywhere (SDK built-ins + user nodes) are collected
      # here in definition order. `Blok::Node.load_user_nodes` snapshots this
      # list before/after requiring each user file and registers only the new
      # subclasses — an explicit, traceable discovery rule (no ObjectSpace scan).
      DESCENDANTS = []

      def self.inherited(subclass)
        super
        DESCENDANTS << subclass
      end

      # Execute the node logic.
      #
      # @param ctx [Blok::Types::Context] The workflow execution context
      # @param config [Hash] Node-specific configuration
      # @return [Object] JSON-serializable result data
      # @raise [NotImplementedError] if not overridden by a subclass
      def execute(ctx, config)
        raise NotImplementedError, "#{self.class}#execute must be implemented"
      end
    end
  end
end
